src/ripple/app/misc/impl/AMMUtils.cpp

```diff
@@ -315,6 +316,7 @@ initializeFeeAuctionVote(
         auctionSlot.setFieldU16(sfDiscountedFee, dfee);
     else if (auctionSlot.isFieldPresent(sfDiscountedFee))
         auctionSlot.makeFieldAbsent(sfDiscountedFee);
+    ammSle->set(&auctionSlot);
```

Comment:
What exactly does this line do that wasn't being done before? It probably needs to also be gated by the amendment.

Reply:
`auctionSlot` was originally fetched via `peekFieldObject()`, which was returning a reference to `auctionSlot`. The latest implementation constructs `auctionSlot` as the local variable, sets its fields, and then sets `auctionSlot` in `AMM` object. `set` originally had declaration only and was not implemented. I don't think it has to be gated.

Reply:
The original code got an existing object and modified it. The new code creates a new object and sets it. So if there are any fields that aren't explicitly set that could have been present when it was retrieved in the previous version with `peekFieldObject` then those fields will be different between the old and new version and that's transaction breaking.

It looks like `sfAuthAccounts` is such a field. This looks transaction breaking without being protected with an amendment to me. If for some reason there's no way `sfAuthAccount` could have ever been set then we'd at the very minimum need a comment, but we'd probably want to rewrite this code to protect that case anyway.

As a side note, is it OK that `sfAuthAccounts` got overwritten with the new code or do we want to preserve that?

Reply:
`sfAuthAccounts` is an optional field and is not going to be present in either the old version or the new version. The final result in terms of which fields are present and which are not is the same in both versions. What changes is how `sfDiscountedFee` is handled when it's fetched from `sfAuctionSlot` because `sfDiscountedFee` is `soeDEFAULT`. The old version fails if the field is not present. The new version fails if the amendment is disabled and succeeds if the amendment is enabled.

Reply:
Just so I understand, `sfAuctionSlot` is guaranteed not exist in `ammSle` on entry to this function? Or it might exist and `sfAuthAccounts` is guaranteed not to exist?

Reply:
Ok. I missed the `peekFieldObject()` change because I was focused on the `set`. I think this is fine, but another option to consider is that you could restore the `peekFieldObject()`, then set up the template here. The only real advantages I can think of offhand are that it would remove the need to create the temporary local, and it avoids a move assignment. The disadvantage is it adds yet another copy of that code block from the ctor. May not be worth it, but I offer for your consideration:
```
    // AMM creator gets the auction slot for free.
    auto& auctionSlot = ammSle->peekFieldObject(sfAuctionSlot);
    if (rules.enabled(fixInnerObjTemplate))
    {
        if (SOTemplate const* elements =
                InnerObjectFormats::getInstance().findSOTemplateBySField(sfAuctionSlot))
            auctionSlot.set(*elements);
    }
    auctionSlot.setAccountID(sfAccount, account);
[etc]
```
But enough bike-shedding.

Reply:
Actually, `sfAuctionSlot` does not exist when called from `AMMCreate` and exists when called from `AMMDeposit`. The latter can be called to restore `AMM` from an empty state. So `sfAuthAccounts` have to be preserved in case it's called from `AMMDeposit` if the amendment is not enabled and reset if the amendment is enabled.

Reply:
How about something like this:
>        STObject auctionSlot = STObject::makeInnerObject(sfAuctionSlot, rules);
>        if (!rules.enabled(fixInnerObjTemplate) && ammSle->isFieldPresent(sfAuctionSlot))
>        {
>            auto& as = ammSle->peekFieldObject(sfAuctionSlot);
>            if (as.isFieldPresent(sfAuthAccounts))
>                auctionSlot.setFieldArray(sfAuthAccounts, as.getFieldArray(sfAuthAccounts));
>        }